contracts/utils/Strings.sol

```diff
@@ -85,10 +84,46 @@ library Strings {
         return toHexString(uint256(uint160(addr)), ADDRESS_LENGTH);
     }
 
+    /**
+     * @dev Converts an `address` with fixed length of 20 bytes to its checksummed ASCII `string` hexadecimal
+     * representation, according to EIP-55.
+     */
+    function toChecksumHexString(address addr) internal pure returns (string memory) {
+        bytes memory lowercase = new bytes(40);
+        uint160 addrValue = uint160(addr);
+        _unsafeSetHexString(lowercase, 0, addrValue);
+        bytes32 hashedAddr = keccak256(abi.encodePacked(lowercase));
+
+        bytes memory buffer = new bytes(42);
+        buffer[0] = "0";
+        buffer[1] = "x";
+        uint160 hashValue = uint160(bytes20(hashedAddr));
+        for (uint256 i = 41; i > 1; --i) {
+            uint8 digit = uint8(addrValue & 0xf);
+            buffer[i] = hashValue & 0xf > 7 ? HEX_DIGITS_UPPERCASE[digit] : HEX_DIGITS[digit];
+            addrValue >>= 4;
+            hashValue >>= 4;
+        }
+        return string(abi.encodePacked(buffer));
+    }
+
     /**
      * @dev Returns true if the two strings are equal.
      */
     function equal(string memory a, string memory b) internal pure returns (bool) {
         return bytes(a).length == bytes(b).length && keccak256(bytes(a)) == keccak256(bytes(b));
     }
+
+    /**
+     * @dev Sets the hexadecimal representation of a value in the specified buffer starting from the given offset.
+     *
+     * NOTE: This function does not check that the `buffer` can allocate `value` without overflowing. Make sure
+     * to check whether `Math.log256(value) + 1` is larger than the specified `length`.
+     */
+    function _unsafeSetHexString(bytes memory buffer, uint256 offset, uint256 value) private pure {
+        for (uint256 i = buffer.length; i > offset; --i) {
+            buffer[i - 1] = HEX_DIGITS[value & 0xf];
```

Comment:
I'm concerned each write may check that `i-1` is within the bound of the array. If that is the case, its a major waste. Maybe we could optimize that.